install/rpm.py: wipe zypper repos before adding new ones [DNM - to be replaced with cherry-pick from upstream]

smithfarm opened the pull request:
Fixes: https://bugzilla.suse.com/show_bug.cgi?id=1151321
Signed-off-by: Nathan Cutler <ncutler@suse.com>

smithfarm commented:
@susebot run teuthology

kshtsk commented:
@susebot run teuthology

kshtsk commented:
@susebot run teuthology

kshtsk commented:
@susebot run teuthology

kshtsk commented:
@susebot run teuthology

kshtsk commented:
@susebot run teuthology

kshtsk commented on teuthology/task/install/rpm.py:
```diff
@@ -132,12 +144,12 @@ def _update_package_list_and_install(ctx, remote, rpm, config):
             raise Exception('Custom repos were specified for %s ' % remote_os +
-                            'but these are currently not supported')
+                            'but these are currently not supported on that OS')
+        _zypper_addrepo(remote, repos)
```
Why this line has been moved here, out of 'opensuse', 'sle' scope? It will fail for other distro.

kshtsk commented on teuthology/task/install/rpm.py:
```diff
@@ -85,6 +85,7 @@ def _zypper_addrepo(remote, repo_list):
     :param repo_list: list of dictionaries with keys 'name', 'url'
     :return:
     """
+    log.debug("Adding zypper repos: %s" % repos)
```
This line cause a failure:
```
2019-09-19T18:13:51.667 ERROR:teuthology.parallel:Exception in parallel execution
Traceback (most recent call last):
  File "/home/ubuntu/teuthology/teuthology/parallel.py", line 83, in __exit__
    for result in self:
  File "/home/ubuntu/teuthology/teuthology/parallel.py", line 101, in next
    resurrect_traceback(result)
  File "/home/ubuntu/teuthology/teuthology/parallel.py", line 19, in capture_traceback
    return func(*args, **kwargs)
  File "/home/ubuntu/teuthology/teuthology/task/install/rpm.py", line 152, in _update_package_list_and_install
    _zypper_addrepo(remote, repos)
  File "/home/ubuntu/teuthology/teuthology/task/install/rpm.py", line 88, in _zypper_addrepo
    log.debug("Adding zypper repos: %s" % repos)
NameError: global name 'repos' is not defined
```

kshtsk replied:
Any reason to duplicate this here?

kshtsk commented:
@susebot run teuthology

kshtsk commented:
@susebot run teuthology

smithfarm commented:
@susebot run teuthology

smithfarm commented:
Looks good to me:

```
2019-09-27T13:32:43.799 DEBUG:teuthology.task.install.rpm:repos before wipe
2019-09-27T13:32:43.799 INFO:teuthology.orchestra.run.target192168000025:Running: 'sudo zypper lr -upEP || true'
2019-09-27T13:32:43.944 INFO:teuthology.orchestra.run.target192168000025.stdout:Warning: No repositories defined.
2019-09-27T13:32:43.945 INFO:teuthology.orchestra.run.target192168000025.stdout:Use the 'zypper addrepo' command to add one or more repositories.
2019-09-27T13:32:43.945 INFO:teuthology.task.install.rpm:Wiping existing repos, if any, because they will be replaced with new ones
2019-09-27T13:32:43.945 INFO:teuthology.orchestra.run.target192168000025:Running: 'sudo rm -f /etc/zypp/repos.d/*'
```

@kshtsk Thanks for the review. I addressed the issues and this has now passed the teuthology deployment test. Do you think it's ready to open a PR upstream?

kshtsk commented on teuthology/task/install/rpm.py:
```diff
@@ -46,8 +46,8 @@ def _remove(ctx, config, remote, rpm):
             _zypper_removerepo(remote, repos)
         else:
-            raise Exception('Custom repos were specified for %s ' % remote_os +
-                            'but these are currently not supported')
```
This change is not related to the issue, so could you please revert it. Besides using `.format` everywhere has no reasonable background and does not help to be python3 compliant.

kshtsk commented on teuthology/task/install/rpm.py:
```diff
@@ -77,14 +77,15 @@ def _package_overrides(pkgs, os):
         result.append(pkg)
     return result
 
-def _zypper_addrepo(remote, repo_list):
```
Please revert this change, it is not related to the issue of this PR and by the way name of signature were selected to correspond zypper long name command for adding repos, i.e. there no `zypper add repos` command, but `zypper addrepo` (long for `zypper ar`.

kshtsk commented on teuthology/task/install/rpm.py:
```diff
@@ -132,12 +144,12 @@ def _update_package_list_and_install(ctx, remote, rpm, config):
+            _zypper_add_repos(remote, repos)
         else:
-            raise Exception('Custom repos were specified for %s ' % remote_os +
-                            'but these are currently not supported')
```
This change is not related to the issue.

kshtsk commented on teuthology/task/install/rpm.py:
```diff
@@ -132,12 +144,12 @@ def _update_package_list_and_install(ctx, remote, rpm, config):
     repos_only = config.get('repos_only')
 
     if repos:
-        log.debug("Adding repos: %s" % repos)
```
This was a useful marker before installing repos.
Maybe it is more convenient to add wiping code before informing user we are trying to add repos?
